In [4]:
import pandas as pd

In [5]:
# Load data
revenue_df = pd.read_csv("sod_revenue.csv")
ad_revenue_df = pd.read_csv("sod_adrevenue.csv")
register_df = pd.read_csv("sod_register.csv")
battle_df = pd.read_csv("sod_battle_short.csv")
session_df = pd.read_csv("sod_session_start.csv")

In [6]:
revenue_df. head()

user_id                  event_time  \
0  f177e97c-4f0e-4820-b202-706f77839fe3  2020-01-23 21:47:22.175000   
1  f177e97c-4f0e-4820-b202-706f77839fe3  2020-01-23 21:48:05.262000   
2  06fc31b3-887b-4309-9ea9-fe2d06471629  2020-01-24 06:53:43.865000   
3  d1f252fc-30e5-4377-af7b-ce98a7f21f43  2020-01-24 16:55:59.090000   
4  ee24a1dc-379d-4fd6-87a5-12969173e103  2020-01-24 18:02:08.691000   

   user_point_in_lifetime user_country user_device_platform  \
0                       8           BR              Android   
1                       8           BR              Android   
2                       4           US                  iOS   
3                      16           US              Android   
4                       0           US              Android   

   user_marketing_user event_type        game_name  \
0                False    revenue  Super Overdrive   
1                False    revenue  Super Overdrive   
2                False    revenue  Super Overdrive   
3                False    revenue  Super Overdrive   
4                 True    revenue  Super Overdrive   

            user_install_time  user_cumulative_iap_revenue  \
0  2020-01-15 20:35:55.409000                         2.97   
1  2020-01-15 20:35:55.409000                         3.96   
2  2020-01-20 06:01:41.207000                         0.99   
3  2020-01-08 00:28:52.021000                         4.99   
4  2020-01-24 01:57:28.926000                         0.99   

   user_cumulative_iap_payments  user_payer_user  event_revenue_gross_usd  \
0                             3             True                     0.99   
1                             4             True                     0.99   
2                             1             True                     0.99   
3                             1             True                     4.99   
4                             1             True                     0.99   

                               event_item_id  user_rank  
0             com.superoverdrive.itemgold.06         17  
1             com.superoverdrive.itemgold.06         17  
2  com.superoverdrive.itembundlespecial.9999          5  
3             com.superoverdrive.itemgold.01          9  
4  com.superoverdrive.itembundlespecial.9999          5

In [23]:
avg_rev= revenue_df['user_cumulative_iap_revenue'].sum()
print((revenue_df['user_id'].nunique()))
print(avg_rev)

5295
1212271.85


In [8]:
ad_revenue_df.head()

user_id                  event_time  \
0  25e77e15-40d6-4d11-a67d-fbb9626b5d33  2020-01-01 01:11:04.647000   
1  5089dc6b-9479-4736-b73b-1f04546ec52a  2020-01-01 01:11:53.981000   
2  5089dc6b-9479-4736-b73b-1f04546ec52a  2020-01-01 01:12:26.402000   
3  df31190f-5b72-4685-974c-f4da44bd133a  2020-01-01 01:12:45.124000   
4  a982313c-a60f-4952-a28b-df3fdf73b01b  2020-01-01 01:12:55.418000   

   user_point_in_lifetime user_country user_device_platform  \
0                     774           US                  iOS   
1                      18           MY              Android   
2                      18           MY              Android   
3                      48           US              Android   
4                     146           CL              Android   

   user_marketing_user  event_type        game_name  \
0                False  ad_revenue  Super Overdrive   
1                False  ad_revenue  Super Overdrive   
2                False  ad_revenue  Super Overdrive   
3                False  ad_revenue  Super Overdrive   
4                False  ad_revenue  Super Overdrive   

            user_install_time  user_cumulative_iap_revenue  \
0  2017-11-17 06:11:31.840000                          0.0   
1  2019-12-13 06:52:08.454000                          0.0   
2  2019-12-13 06:52:08.454000                          0.0   
3  2019-11-13 18:27:28.745000                          0.0   
4  2019-08-07 20:17:36.781000                          0.0   

   user_cumulative_iap_payments  user_payer_user  event_revenue_gross_usd  \
0                             0            False                   0.0015   
1                             0            False                   0.0024   
2                             0            False                   0.0024   
3                             0            False                   0.0055   
4                             0            False                   0.0016   

   user_rank  
0         30  
1         19  
2         19  
3         30  
4         14

In [9]:
battle_df.head()

user_id                  event_time  \
0  9b685c09-524a-43e2-a6f9-b970a438fd33  2020-01-01 01:47:39.837000   
1  2b31d675-2007-4689-9f0c-cd67bb3e04c3  2020-01-01 01:47:45.451000   
2  139b74ef-3178-42c1-a4b5-6f8e8b490af3  2020-01-01 01:48:52.189000   
3  044d0f29-97b7-4abe-874c-38161bc5cf29  2020-01-01 01:48:59.888000   
4  57e63954-a7ca-4af1-9ae4-a4aed0f89334  2020-01-01 01:49:15.450000   

   user_point_in_lifetime user_country user_device_platform  \
0                       2           RU              Android   
1                       0           VE              Android   
2                      46           US              Android   
3                       0           MX              Android   
4                       1           US              Android   

   user_marketing_user event_type        game_name  \
0                False     battle  Super Overdrive   
1                False     battle  Super Overdrive   
2                False     battle  Super Overdrive   
3                False     battle  Super Overdrive   
4                False     battle  Super Overdrive   

            user_install_time  user_cumulative_iap_revenue  \
0  2019-12-29 08:38:21.177000                          0.0   
1  2019-12-31 20:00:28.676000                          0.0   
2  2019-11-15 15:56:15.092000                          0.0   
3  2020-01-01 00:58:12.397000                          0.0   
4  2019-12-30 04:45:54.467000                          0.0   

   user_cumulative_iap_payments  user_payer_user event_battle_result  \
0                             0            False          2-Star Win   
1                             0            False          3-Star Win   
2                             0            False          3-Star Win   
3                             0            False          2-Star Win   
4                             0            False          2-Star Win   

  custom_battle_type  event_battle_duration  user_cumulative_battles  \
0           campaign                    141                      138   
1           campaign                    177                       10   
2           campaign                    200                     1398   
3           campaign                    173                       12   
4           campaign                    215                       16   

   custom_map_id  user_rank  
0            203         11  
1            205          3  
2            505         30  
3            204          2  
4            206          4

In [7]:
# Preprocess revenue Data


# Convert to datetime
revenue_df['event_time'] = pd.to_datetime(revenue_df['event_time'])
revenue_df['user_install_time'] = pd.to_datetime(revenue_df['user_install_time'])

# Extract event date
revenue_df['event_date'] = revenue_df['event_time'].dt.date


# Daily revenue split by platform and country


In [8]:

# Group by date, platform, country
revenue_by_platform_country = revenue_df.groupby(
    ['event_date', 'user_device_platform', 'user_country']
)['event_revenue_gross_usd'].sum().reset_index()

revenue_by_platform_country.rename(columns={'event_revenue_gross_usd': 'daily_revenue'}, inplace=True)

print(revenue_by_platform_country)


      event_date user_device_platform user_country  daily_revenue
0     2020-01-01              Android           AT           0.99
1     2020-01-01              Android           AU           4.95
2     2020-01-01              Android           BE           9.98
3     2020-01-01              Android           BH           0.99
4     2020-01-01              Android           BR           2.97
...          ...                  ...          ...            ...
1900  2020-01-31                  iOS           TH           9.99
1901  2020-01-31                  iOS           TW         369.90
1902  2020-01-31                  iOS           UA          13.95
1903  2020-01-31                  iOS           US        1026.22
1904  2020-01-31                  iOS           VN           2.97

[1905 rows x 4 columns]


In [34]:
revenue_by_platform_country.to_csv("revenue_by_platform_country.csv", index=False)

print("revenue_by_platform_country.csv")

revenue_by_platform_country.csv


#  2. KPI 1: ARPPU Over Time (Revenue / Paying Users)

In [9]:
# Filter paying users
payers = revenue_df[revenue_df['user_payer_user'] == True]

# Group by date
daily_arppu = payers.groupby('event_date').agg({
    'event_revenue_gross_usd': 'sum',
    'user_id': 'nunique'
}).reset_index()

daily_arppu['arppu'] = daily_arppu['event_revenue_gross_usd'] / daily_arppu['user_id']

print(daily_arppu)


    event_date  event_revenue_gross_usd  user_id      arppu
0   2020-01-01                  4545.91      283  16.063286
1   2020-01-02                  3602.21      274  13.146752
2   2020-01-03                  4185.49      299  13.998294
3   2020-01-04                  4639.61      321  14.453614
4   2020-01-05                  4142.82      296  13.996014
5   2020-01-06                  2542.91      222  11.454550
6   2020-01-07                  1596.32      193   8.271088
7   2020-01-08                  2140.23      209  10.240335
8   2020-01-09                  2027.93      221   9.176154
9   2020-01-10                  4340.01      329  13.191520
10  2020-01-11                  4564.81      346  13.193092
11  2020-01-12                  4377.51      371  11.799218
12  2020-01-13                  2791.01      265  10.532113
13  2020-01-14                  2946.64      235  12.538894
14  2020-01-15                  2979.42      247  12.062429
15  2020-01-16                  2735.35 

In [11]:
avg_arppu = daily_arppu['arppu'].mean()
print(avg_arppu)


12.337677019430005


In [35]:
daily_arppu.to_csv("daily_arppu.csv", index=False)

print("daily_arppu.csv")

daily_arppu.csv


# KPI's : Installs, LTV, DAU, MAU

In [1]:
import pandas as pd

# Load data with proper datetime parsing
session_df = pd.read_csv("sod_session_start.csv")
register_df = pd.read_csv("sod_register.csv")
revenue_df = pd.read_csv("sod_revenue.csv")

In [3]:
# Explicitly convert event_time columns to datetime
session_df['event_time'] = pd.to_datetime(session_df['event_time'], errors='coerce')
register_df['event_time'] = pd.to_datetime(register_df['event_time'], errors='coerce')
revenue_df['event_time'] = pd.to_datetime(revenue_df['event_time'], errors='coerce')

user_id              event_time  \
0      f177e97c-4f0e-4820-b202-706f77839fe3 2020-01-23 21:47:22.175   
1      f177e97c-4f0e-4820-b202-706f77839fe3 2020-01-23 21:48:05.262   
2      06fc31b3-887b-4309-9ea9-fe2d06471629 2020-01-24 06:53:43.865   
3      d1f252fc-30e5-4377-af7b-ce98a7f21f43 2020-01-24 16:55:59.090   
4      ee24a1dc-379d-4fd6-87a5-12969173e103 2020-01-24 18:02:08.691   
...                                     ...                     ...   
12358  ce009465-7f8a-4232-b1d4-8e4d51c569d4 2020-01-14 13:35:36.286   
12359  d7f64e4a-c06d-4b72-ab08-ac7396b02751 2020-01-14 16:20:23.853   
12360  d7f64e4a-c06d-4b72-ab08-ac7396b02751 2020-01-14 16:37:13.132   
12361  e0005ced-4260-4c06-9710-606812170b54 2020-01-14 17:23:27.768   
12362  4883bf19-2436-4b2e-87e5-8b1b904e537a 2020-01-14 18:44:33.271   

       user_point_in_lifetime user_country user_device_platform  \
0                           8           BR              Android   
1                           8           BR              Android   
2                           4           US                  iOS   
3                          16           US              Android   
4                           0           US              Android   
...                       ...          ...                  ...   
12358                       5           BR              Android   
12359                     877           US              Android   
12360                     877           US              Android   
12361                       3           IT                  iOS   
12362                       0           SK              Android   

       user_marketing_user event_type        game_name  \
0                    False    revenue  Super Overdrive   
1                    False    revenue  Super Overdrive   
2                    False    revenue  Super Overdrive   
3                    False    revenue  Super Overdrive   
4                     True    revenue  Super Overdrive   
...                    ...        ...              ...   
12358                False    revenue  Super Overdrive   
12359                False    revenue  Super Overdrive   
12360                False    revenue  Super Overdrive   
12361                 True    revenue  Super Overdrive   
12362                False    revenue  Super Overdrive   

                user_install_time  user_cumulative_iap_revenue  \
0      2020-01-15 20:35:55.409000                         2.97   
1      2020-01-15 20:35:55.409000                         3.96   
2      2020-01-20 06:01:41.207000                         0.99   
3      2020-01-08 00:28:52.021000                         4.99   
4      2020-01-24 01:57:28.926000                         0.99   
...                           ...                          ...   
12358  2020-01-08 19:32:39.475000                         0.99   
12359  2017-08-19 17:32:21.658000                       119.93   
12360  2017-08-19 17:32:21.658000                       139.92   
12361  2020-01-11 13:10:41.844000                        19.98   
12362  2020-01-14 07:58:38.525000                         0.99   

       user_cumulative_iap_payments  user_payer_user  event_revenue_gross_usd  \
0                                 3             True                     0.99   
1                                 4             True                     0.99   
2                                 1             True                     0.99   
3                                 1             True                     4.99   
4                                 1             True                     0.99   
...                             ...              ...                      ...   
12358                             1             True                     0.99   
12359                             7             True                    19.99   
12360                             8             True                    19.99   
12361                             2             True                     9.99 

In [4]:
# Add date and month columns
session_df['date'] = session_df['event_time'].dt.date
session_df['month'] = session_df['event_time'].dt.to_period('M')

register_df['date'] = register_df['event_time'].dt.date

revenue_df['date'] = revenue_df['event_time'].dt.date
revenue_df['month'] = revenue_df['event_time'].dt.to_period('M')


In [5]:
# Daily Active Users
dau = session_df.groupby('date')['user_id'].nunique().reset_index(name='DAU')

# MAU
mau = session_df.groupby('month')['user_id'].nunique().reset_index(name='MAU')

# Installs per day
installs = register_df.groupby('date')['user_id'].nunique().reset_index(name='Installs')

# Revenue per day
daily_revenue = revenue_df.groupby('date')['event_revenue_gross_usd'].sum().reset_index(name='Revenue')

# LTV = Revenue / Installs (per day)
ltv_df = pd.merge(installs, daily_revenue, on='date', how='left')
ltv_df['LTV'] = ltv_df['Revenue'] / ltv_df['Installs']
ltv_df.fillna(0, inplace=True)


In [8]:
print(dau)
print(mau)

          date    DAU
0   2020-01-01  24469
1   2020-01-02  22711
2   2020-01-03  23198
3   2020-01-04  24515
4   2020-01-05  25023
5   2020-01-06  23428
6   2020-01-07  23229
7   2020-01-08  23675
8   2020-01-09  24707
9   2020-01-10  25369
10  2020-01-11  27379
11  2020-01-12  28005
12  2020-01-13  25881
13  2020-01-14  25682
14  2020-01-15  25953
15  2020-01-16  25983
16  2020-01-17  25338
17  2020-01-18  26510
18  2020-01-19  27395
19  2020-01-20  25740
20  2020-01-21  25241
21  2020-01-22  25743
22  2020-01-23  25920
23  2020-01-24  28085
24  2020-01-25  29100
25  2020-01-26  29990
26  2020-01-27  27146
27  2020-01-28  26936
28  2020-01-29  26890
29  2020-01-30  26681
30  2020-01-31  27732
     month     MAU
0  2020-01  262508


In [10]:
print(installs)
print(ltv_df)

          date  Installs
0   2020-01-01      5007
1   2020-01-02      4841
2   2020-01-03      5048
3   2020-01-04      5753
4   2020-01-05      5919
5   2020-01-06      4793
6   2020-01-07      4812
7   2020-01-08      5208
8   2020-01-09      5704
9   2020-01-10      5969
10  2020-01-11      6900
11  2020-01-12      7025
12  2020-01-13      5697
13  2020-01-14      5493
14  2020-01-15      5389
15  2020-01-16      5368
16  2020-01-17      5549
17  2020-01-18      6660
18  2020-01-19      6745
19  2020-01-20      5734
20  2020-01-21      5400
21  2020-01-22      5727
22  2020-01-23      5792
23  2020-01-24      6149
24  2020-01-25      7194
25  2020-01-26      7894
26  2020-01-27      5971
27  2020-01-28      5801
28  2020-01-29      5885
29  2020-01-30      5925
30  2020-01-31      6542
          date  Installs  Revenue       LTV
0   2020-01-01      5007  4545.91  0.907911
1   2020-01-02      4841  3602.21  0.744105
2   2020-01-03      5048  4184.50  0.828942
3   2020-01-04      5753

In [11]:

# Combine DAU, Installs, Revenue, LTV for export

combined = pd.merge(dau, installs, on='date', how='outer')

combined

date    DAU  Installs
0   2020-01-01  24469      5007
1   2020-01-02  22711      4841
2   2020-01-03  23198      5048
3   2020-01-04  24515      5753
4   2020-01-05  25023      5919
5   2020-01-06  23428      4793
6   2020-01-07  23229      4812
7   2020-01-08  23675      5208
8   2020-01-09  24707      5704
9   2020-01-10  25369      5969
10  2020-01-11  27379      6900
11  2020-01-12  28005      7025
12  2020-01-13  25881      5697
13  2020-01-14  25682      5493
14  2020-01-15  25953      5389
15  2020-01-16  25983      5368
16  2020-01-17  25338      5549
17  2020-01-18  26510      6660
18  2020-01-19  27395      6745
19  2020-01-20  25740      5734
20  2020-01-21  25241      5400
21  2020-01-22  25743      5727
22  2020-01-23  25920      5792
23  2020-01-24  28085      6149
24  2020-01-25  29100      7194
25  2020-01-26  29990      7894
26  2020-01-27  27146      5971
27  2020-01-28  26936      5801
28  2020-01-29  26890      5885
29  2020-01-30  26681      5925
30  2020-01-31  27732      6542

In [ ]:
combined = pd.merge(combined, ltv_df[['date', 'Revenue', 'LTV']], on='date', how='outer')

combined = combined.sort_values(by='date')

In [29]:
# Export

combined.to_csv("dau_installs_ltv.csv", index=False)

mau.to_csv("mau.csv", index=False)

In [30]:
print("dau_installs_ltv.csv")

dau_installs_ltv.csv


In [31]:
print("mau.csv")

mau.csv


In [27]:
print (combined)

          date    DAU  Installs  Revenue       LTV
0   2020-01-01  24469      5007  4545.91  0.907911
1   2020-01-02  22711      4841  3602.21  0.744105
2   2020-01-03  23198      5048  4184.50  0.828942
3   2020-01-04  24515      5753  4629.62  0.804731
4   2020-01-05  25023      5919  4142.82  0.699919
5   2020-01-06  23428      4793  2541.92  0.530340
6   2020-01-07  23229      4812  1596.32  0.331737
7   2020-01-08  23675      5208  2140.23  0.410950
8   2020-01-09  24707      5704  2027.93  0.355528
9   2020-01-10  25369      5969  4320.02  0.723743
10  2020-01-11  27379      6900  4564.81  0.661567
11  2020-01-12  28005      7025  4377.51  0.623133
12  2020-01-13  25881      5697  2791.01  0.489909
13  2020-01-14  25682      5493  2946.64  0.536435
14  2020-01-15  25953      5389  2979.42  0.552871
15  2020-01-16  25983      5368  2735.35  0.509566
16  2020-01-17  25338      5549  3287.83  0.592509
17  2020-01-18  26510      6660  3423.98  0.514111
18  2020-01-19  27395      6745

In [28]:
print(mau)

     month     MAU
0  2020-01  262508


# Bonus: Recommendation for Campaign Battles

In [43]:
battle_df = pd.read_csv("sod_battle_short.csv")
campaign_battles = battle_df[battle_df['custom_battle_type'] == 'campaign']

# Calculate average revenue of the users

avg_revenue_campaign = battle_df['user_cumulative_iap_revenue'].sum()/battle_df['user_id'].nunique()
print("Average revenue from campaign players:", avg_revenue_campaign)

Average revenue from campaign players: 419.56045330693087


In [45]:
print(campaign_battles['user_id'].nunique())

122213


In [44]:
common_user_id_rev = revenue_df[revenue_df['user_id'].isin(battle_df['user_id'])]
print(common_user_id_rev['user_id'].nunique())

5200


In [47]:
avg_revenue_not_campaign = common_user_id_rev['user_cumulative_iap_revenue'].sum()/common_user_id_rev['user_id'].nunique()
print("Average revenue from non campaign players:", avg_revenue_not_campaign)

Average revenue from non campaign players: 231.75578269230766


In [41]:
battle_df['user_cumulative_iap_revenue'].sum()/battle_df['user_id'].nunique()

419.56045330693087

In [28]:
avg_rev_without_campaign = common_user_id_rev ['user_cumulative_iap_revenue'].sum()/common_user_id_rev['user_id'].nunique()
print(avg_rev_without_campaign)

231.75578269230766


Average Revenue per Campaign Player:<span style='color:Green'>  $ 419.56 </span>

Average Revenue per non Campaign Player: <span style='color:Red'>$ 231.75 </span>

Campaign players spend ~ 81% more than non-campaign players.

Interpretation: Players who participate in campaign battles tend to generate higher revenue than average players.

This suggests that campaign mode attracts high-value users or motivates spending (possibly due to progression mechanics, difficulty spikes, or story engagement).

Recommendation: Double down on campaign features: 
- Add special offers
- exclusive rewards
- or difficulty-based monetization (e.g., retry tokens, boosters)
- Can also consider targeted push notifications or promotions for players who show interest in campaign mode.